src/complexity_measures/statistical_complexity.jl

```diff
@@ -0,0 +1,197 @@
+            dist::SemiMetric, est::ProbabilitiesEstimator,
+            entr::EntropyDefinition)
+        StatisticalComplexity(dist, est, entr)
+    end
+end
+
+function complexity(c::StatisticalComplexity, x::AbstractVector{T}) where T
+    (; dist, est, entr) = c
+
+    p = probabilities(est, x)
+
+    H_q = entropy_normalized(entr, est, x)
+
+    L = total_outcomes(est, x)
+
+    # calculate distance between calculated distribution and uniform one
+    D_q = evaluate(dist, p.p, fill(1.0/L, size(p)))
+
+    # generate distribution with just one filled bin
+    deterministic = zeros(size(p))
+    deterministic[1] = 1
+
+    D_max = evaluate(dist, deterministic, fill(1.0/L, size(p)))
+    C_q = D_q / D_max * H_q
+
+    return C_q
+end
+
+function complexity(c::StatisticalComplexity, p::Probabilities)
+    (; dist, est, entr) = c
+    if !(est isa SymbolicPermutation || est isa SymbolicWeightedPermutation || est isa SymbolicAmplitudeAwarePermutation)
+        throw(
+            ArgumentError(
+                "To estimate the statistical complexity directly from a probabiliy distribution, "*
+                "the probabilities estimator must be one of"*
+                "`SymbolicPermutation`, `SymbolicAmplitudeAwarePermutation` or `SymbolicWeightedPermutation`"
+            )
+        )
+    end
+    L = total_outcomes(est, randn(100))
+    norm = log(entr.base, L)
+    H_q = entropy(entr, p) / norm
+
+    # calculate distance between calculated distribution and uniform one
+    D_q = evaluate(dist, p.p, fill(1.0/L, size(p)))
+
+    # generate distribution with just one filled bin
+    deterministic = zeros(size(p))
+    deterministic[1] = 1
+
+    D_max = evaluate(dist, deterministic, fill(1.0/L, size(p)))
+    C_q = D_q / D_max * H_q
+
+    return C_q
+end
+
+function fill_probs_k!(p, prob_params, L, i, k)
+    p.p[1] = prob_params[k] # why set first element here if overwriting it in the loop below?
+    # if we know that p has sufficient length relative to L and i,
+    # @inbounds can save some computation time by skipping bounds checking.
+    @inbounds for j = 1:(L - i) 
+        p.p[j] = (1 - prob_params[k]) / (L - i)
+    end
+end
+
+linearpermissiverange(start; stop, length) = length==1 ? (start:start) : range(start, stop=stop, length=length)
+
+"""
+    maximum_complexity_entropy(c::StatisticalComplexity; num=1)
+Calculate the maximum complexity-entropy curve for the statistical complexity according to [^Rosso2007]
+for `num * total_outcomes(c.est)` different values of the normalized permutation entropy.
+
+## Description
+The way the statistical complexity is designed, there is a minimum and maximum possible complexity
+for data with a given permutation entropy.
+The calculation time of the maximum complexity curve grows as `O(total_outcomes(c.est)^2)`, and thus takes
+very long for high numbers of outcomes.
+This function is adapted from S. Sippels implementation in statcomp [^statcomp].
```

Comment:
This is tricky. But I think we are safe. We are implementing the algorithm that has mathematical equations and we can't change them. Just because someone implemented these equations first in some other language does it mean that we have to get its license ? This doesn't make sense to me.

Reply:
> Just because someone implemented these equations first in some other language does it mean that we have to get its license ? 

I'm playing devil's advocate here @Datseris,  because it explicitly says in the docstring that this work is _adapted_ from some GPL-2 software, which may imply that the original source code has been copied, then modified somehow.  That can potentially lead to issues, because GPL-2 is a copyleft, meaning that any derived source code must be licensed under the same conditions as the original code. [Here's](https://www.tldrlegal.com/license/gnu-general-public-license-v2) a summary of GPL-2. Another summary is [here](https://choosealicense.com/licenses/gpl-2.0/). Note especially the parts under "conditions" in the latter link.

That said, I've downloaded the statcomp package and investigated the source code for relevant functions, and I don't think this in any way can be classified as an *adaptation* of some other code. In fact, I can't really find any trace of  adapted code, as far as I can see. We use our own API to do most of the work, and what remains are, well, implementations of equations from the paper. So I, too, think we're safe here. 

Could you just clarify which parts of the code you adapted, @ikottlarz? Or did you mean that you just used the R code to get an understanding of the equations in the paper?  An attribution to the author of the R code is well-earned in any case if you found it useful for this implementation, even though this code looks nothing like the R code. Since the docstring gives credit already, I think this should be just fine.

Reply:
I agree with what @kahaaga said, so as long as the docstring is rephrased we are in the clear.